ir/memory.cpp

```diff
@@ -431,6 +431,39 @@ Memory::Memory(State &state, bool little_endian)
   local_block_liveness
     = expr::mkLambda({ expr::mkVar("#bid0", bitsBid() - 1) }, false);
 
+  // The first byte of the memory cannot be allocated because it is for a null
+  // pointer.
+  // The last byte of the memory cannot be allocated because ptr + size
+  // (where size is the size of the block and ptr is the beginning address of
+  // the block) should not overflow.
+  avail_space = expr::mkVar("avail_space", bitsPtrSize());
+  if (state.isSource()) {
+    state.addAxiom(avail_space.ult(expr::mkUInt(-2, bitsPtrSize())));
+  }
+
+  // TODO: replace the magic number 2 with the result of analysis.
+  // This is just set as 2 to make existing unit tests run without timeout.
+  unsigned non_local_bid_upperbound = 2;
+  // Maximum alignment that a non-local block can have
+  // TODO: get this from data layout?
+  auto max_align_minus1 = expr::mkUInt(8 - 1, bitsPtrSize());
+
+  // Omit null pointer
+  for (unsigned bid = 1; bid <= non_local_bid_upperbound; ++bid) {
+    Pointer p(*this, bid, false);
+
+    // The required space size should consider alignment
+    auto size_upperbound = p.block_size() + max_align_minus1;
```

Comment:
Maximum alignment was used here because memory does not store alignment per block - do you think we should store them?

Reply:
Right, we only have this information for the allocated globals, not for other non-local stuff.
OK, let's keep this maximum alignment; it's not worth it to complicate further the encoding.
But move this to a field in memory.h so it can be filled out by Target data.